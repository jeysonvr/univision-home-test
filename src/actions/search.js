import { types } from "../types/types";

export const setKeyword = (keyword, offset) => {
    
    return (dispatch) => {
        dispatch( {
            type: types.setParameters,
            payload: {
                keyword: keyword,
                offset: offset
            }
        });
    }
}

export const setFullImage = (url, title) => {
    
    return (dispatch) => {
        dispatch( {
            type: types.setFullImage,
            payload: {
                imgURL: url,
                imgTitle: title
            }
        });
    }
}

export const setScrollSlider = (scroll) => {

    return (dispatch) => {
        dispatch( {
            type: types.setScrollSlider,
            payload: {
                scroll: scroll
            }
        });
    }

}