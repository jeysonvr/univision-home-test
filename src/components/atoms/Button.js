import React from 'react';

export const Button = ({ handleClick, icon, classList='', content='' }) => {
    return (
        <>
            <button 
                className={ classList } 
                onClick={ handleClick }
            >
                <i className={ icon }></i>
                { content }
            </button>
        </>
    )
}
