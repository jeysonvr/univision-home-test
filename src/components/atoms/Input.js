import React from 'react'

export const Input = ({ keyword, handleInputChange }) => {
    return (
        <input 
            className="col-12 col-sm-10 col-md-6"
            style={{marginRight: '10px'}}
            type="text"
            placeholder="Search ..."
            autoComplete="off"
            name="keyword"
            value={ keyword }
            onChange={ handleInputChange }
        />
    )
}
