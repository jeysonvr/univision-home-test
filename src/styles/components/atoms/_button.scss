.button__close {
    background: transparent;
    position: absolute;
    right: -15px;
    top: -15px;
}