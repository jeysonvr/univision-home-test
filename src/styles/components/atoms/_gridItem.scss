.gridItem__card-grid {
    margin: 5px 30px;
    scroll-behavior: smooth;
    overflow-x: scroll;
    overflow-y: hidden;
    white-space: nowrap;
    display: flex;
    align-items: center;

    & img, div {
        display: flex;
        zoom: 1;
        max-height: 170px;
    }

    &.fast{
        scroll-behavior: unset;
    }
}

.gridItem__container {
    margin: 5px;

    & img {
        max-width: unset;
        width: 150px;
        cursor: pointer;
    }
}

.gridItem__button {
    height: 100vh;
    position: absolute;
    z-index: 1;

    & button {
        background: gray;
        border: none;
        border-radius: 50%;
        color: white;
        margin: auto;
        width: 40px;
        height: 40px;
        opacity: 0.5;

        &:hover {
            background: rgb(46, 46, 46);
            transition: opacity 0.3s ease;
            opacity: 1;
        }
    }

    &.left {
        left: 20px;
    }

    &.right {
        right: 20px;
    }

}

::-webkit-scrollbar {
    width: 0;  /* Remove scrollbar space */
    background: transparent;  /* Optional: just make scrollbar invisible */
}