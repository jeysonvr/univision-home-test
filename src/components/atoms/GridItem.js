import React, { useEffect, useRef, useState } from 'react';
import { useDispatch } from "react-redux";
import { setFullImage } from '../../actions/search';
import { elementInViewport } from '../../helpers/extra';

export const GridItem = ({id, title, url, parentScrollPosition = 0}) => {

    const inputRef = useRef();
    const [path, setPath] = useState('');
    const dispatch = useDispatch();
    

    useEffect(() => {
        if( !path && elementInViewport(inputRef.current, parentScrollPosition) ) {
            setPath(url);
        }
    }, [parentScrollPosition]);

    const handleImageClick = () => {
        dispatch( setFullImage(url, title) );
    }

    return (
        <div className="gridItem__container animate__animated animate__fadeIn">
            <img 
                alt={ title } 
                className="img-thumbnail lazy"
                onClick={ () => handleImageClick(url, title) } 
                ref={ inputRef }
                src={ path } 
            />
        </div>
    )
}
