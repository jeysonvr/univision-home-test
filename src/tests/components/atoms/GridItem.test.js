import React from 'react';
import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import { GridItem } from '../../../components/atoms/GridItem';
import { setFullImage } from '../../../actions/search';

jest.mock('../../../actions/search', () => ({
    setFullImage: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
    search: {
        keyword: 'test', 
        offset: 0, 
        imgURL: 'a', 
        imgTitle: 'b'
    }
};
let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <GridItem />
    </Provider>
);

describe('Test in <GridItem />', () => {

    beforeEach( () => {
        store = mockStore(initialState);
        jest.clearAllMocks();
    })

    test('should show component correctly', () => {
        expect( wrapper ).toMatchSnapshot(); 
    })

    test('should trigger action handleImageClick', () => {
        wrapper.find('img').prop('onClick')();
        expect( setFullImage ).toHaveBeenCalled();
    })

})