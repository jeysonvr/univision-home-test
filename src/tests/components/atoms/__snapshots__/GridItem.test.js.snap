// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <GridItem /> should show component correctly 1`] = `
<div
  className="gridItem__container animate__animated animate__fadeIn"
>
  <img
    className="img-thumbnail lazy"
    onClick={[Function]}
  />
</div>
`;
